crates/swc_ecma_parser/src/lexer/mod.rs

```diff
@@ -0,0 +1,348 @@
+//! ECMAScript lexer.
+
+use std::{cell::RefCell, char, iter::FusedIterator, rc::Rc};
+
+use swc_atoms::AtomStoreCell;
+use swc_common::{
+    comments::Comments,
+    input::{Input, StringInput},
+    BytePos, Span,
+};
+use swc_ecma_ast::EsVersion;
+use swc_ecma_lexer::common::lexer::{
+    char::CharExt, comments_buffer::CommentsBuffer, fixed_len_span, pos_span, LexResult,
+    Lexer as LexerTrait,
+};
+
+use self::table::{ByteHandler, BYTE_HANDLERS};
+use crate::{
+    error::{Error, SyntaxError},
+    Context, Syntax,
+};
+
+mod jsx;
+mod state;
+mod table;
+mod token;
+pub mod util;
+
+pub(crate) use token::{NextTokenAndSpan, Token, TokenAndSpan, TokenValue};
+
+#[derive(Clone)]
+pub struct Lexer<'a> {
+    comments: Option<&'a dyn Comments>,
+    /// [Some] if comment comment parsing is enabled. Otherwise [None]
+    comments_buffer: Option<CommentsBuffer>,
+
+    pub ctx: Context,
+    input: StringInput<'a>,
+    start_pos: BytePos,
+
+    state: self::state::State,
+    pub(crate) syntax: Syntax,
+    pub(crate) target: EsVersion,
+
+    errors: Rc<RefCell<Vec<Error>>>,
+    module_errors: Rc<RefCell<Vec<Error>>>,
+
+    buf: Rc<RefCell<String>>,
+
+    atoms: Rc<AtomStoreCell>,
+}
+
+impl FusedIterator for Lexer<'_> {}
+
+impl<'a> swc_ecma_lexer::common::lexer::Lexer<'a, TokenAndSpan> for Lexer<'a> {
+    type State = self::state::State;
+    type Token = self::Token;
+
+    #[inline(always)]
+    fn input(&self) -> &StringInput<'a> {
+        &self.input
+    }
+
+    #[inline(always)]
+    fn input_mut(&mut self) -> &mut StringInput<'a> {
+        &mut self.input
+    }
+
+    #[inline(always)]
+    fn push_error(&self, error: Error) {
+        self.errors.borrow_mut().push(error);
+    }
+
+    #[inline(always)]
+    fn state(&self) -> &Self::State {
+        &self.state
+    }
+
+    #[inline(always)]
+    fn state_mut(&mut self) -> &mut Self::State {
+        &mut self.state
+    }
+
+    #[inline(always)]
+    fn comments(&self) -> Option<&'a dyn swc_common::comments::Comments> {
+        self.comments
+    }
+
+    #[inline(always)]
+    fn comments_buffer(
+        &self,
+    ) -> Option<&swc_ecma_lexer::common::lexer::comments_buffer::CommentsBuffer> {
+        self.comments_buffer.as_ref()
+    }
+
+    #[inline(always)]
+    fn comments_buffer_mut(
+        &mut self,
+    ) -> Option<&mut swc_ecma_lexer::common::lexer::comments_buffer::CommentsBuffer> {
+        self.comments_buffer.as_mut()
+    }
+
+    #[inline(always)]
+    unsafe fn input_slice(&mut self, start: BytePos, end: BytePos) -> &'a str {
+        let s = self.input.slice(start, end);
+        // SAFETY: the input is guaranteed to live for the entire
+        // lifetime.
+        std::mem::transmute::<&str, &'a str>(s)
+    }
+
+    fn input_uncons_while(&mut self, f: impl FnMut(char) -> bool) -> &'a str {
```

Comment:
Can't we bound `&mut self` instead? This feels like an unnecessary transmute

Reply:
The latest commit resolve this issue, though it introduces changes.